qt/scientific_interfaces/EnggDiffraction/RunMap.h

```diff
@@ -0,0 +1,52 @@
+#ifndef MANTIDQTCUSTOMINTERFACES_ENGGDIFFRACTION_RUNMAP_H_
+#define MANTIDQTCUSTOMINTERFACES_ENGGDIFFRACTION_RUNMAP_H_
+
+#include <array>
+#include <unordered_map>
+#include <vector>
+
+namespace MantidQt {
+namespace CustomInterfaces {
+
+/**
+Templated container class holding information relating to runs, indexed by
+run number and bank ID.
+*/
+template <size_t NumBanks, typename T> class RunMap {
+
+public:
+  /**
+   Add an item to the map
+   @param runNumber The run number associated with the item
+   @param bank The bank ID associated with the item
+   @param itemToAdd The item to add
+   */
+  void add(const int runNumber, const size_t bank, const T itemToAdd);
```

Comment:
How large items are going to be added? Would `const T &itemToAdd` make any sense?

Reply:
They are almost always shared pointers to workspaces (such as [here](https://github.com/mantidproject/mantid/blob/master/qt/scientific_interfaces/EnggDiffraction/EnggDiffFittingModel.h#L83)). Is it considered better to pass these by value or reference?

Reply:
If passed by reference, one has to make sure the reference doesn't outlive the actual pointer. I think a reference would be safe here unless you are calling `add()` in some obscure asynchronous way. The possible performance impact for copying shared pointers is not a big deal, though, so if you feel like staying on the safe side, you can leave the method as-is.